### PR TITLE
Tenancy for queries

### PR DESCRIPTION
**Pull request by esnible**
Signed-off-by: Ed Snible <snible@us.ibm.com>

This PR replaces https://github.com/jaegertracing/jaeger/pull/3735

- This PR flows the tenant header through the query APIs.
- This PR blocks requests with invalid tenant or missing tenant header from reaching the query APIs.
- This PR does **NOT** do authorization.  It is assumed authorization is handled by a reverse proxy or other non-Jaeger component at this time.
- This PR requires actual distinguishing of the tenant to be done by the storage layer.  The purpose of this PR is to be sure the storage layer receives a tenant that is valid when Jaeger is started with `--multi_tenant.enabled=true`.

cc @pavolloffay

### CODE REVIEW

File: cmd/query/app/apiv3/grpc_gateway.go
Comment (pavolloffay) [inline cmd/query/app/apiv3/grpc_gateway.go:37]: why is this hardcoded? It deserves a comment

Comment (pavolloffay) [reply cmd/query/app/apiv3/grpc_gateway.go:37]: making this API more configurable seems better. The function has also a lot of parameters, maybe having a params struct would simplify it as well.

File: pkg/config/tenancy/context.go
Comment (esnible) [reply pkg/config/tenancy/context.go:36]: What values are you thinking?  Typically PermissionDenied messages are cryptic, as they often get sent to the client, who is only supposed to know enough to fix the problem, but not enough to guess how to get around the security.

File: pkg/config/tenancy/context.go
Comment (esnible) [reply pkg/config/tenancy/context.go:77]: The /v3 HTTP API is implemented by calling the gRPC implementation.  I believe you did this work.  MetadataAnnotator() is a new middleware that will be called once per v3 query.

I believe github.com/grpc-ecosystem/grpc-gateway, or possibly Go's GRPC itself, takes care of merging all gRPC metadata together, so there shouldn't be a concern about other values.

If you are concerned about a `map` allocation per query these could be cached by the middleware.

File: pkg/config/tenancy/context.go
Comment (pavolloffay) [reply pkg/config/tenancy/context.go:36]: We could put there the list of tenants

Comment (esnible) [reply pkg/config/tenancy/context.go:36]: That's a security no-no.  We can't say 'You identified as tenant "dummy", to use this site you must represent yourself as "ibm", "redhat", or "google".'  If we give any information as part of PermissionDenied we will fail any kind of security audit or pen test.

**Review by pavolloffay — APPROVED** [review]

File: pkg/config/tenancy/interceptor.go
Comment (yurishkuro) [inline pkg/config/tenancy/interceptor.go:38]:
```suggestion
// NewGuardingStreamInterceptor blocks handling of streams whose tenancy header doesn't meet tenancy requirements.
```

File: pkg/config/tenancy/tenancy.go
Comment (yurishkuro) [inline pkg/config/tenancy/tenancy.go:37]: this file would be better named `config.go`

File: pkg/config/tenancy/context.go
Comment (yurishkuro) [inline pkg/config/tenancy/context.go:75]: this feels dangerous, isn't it possible to have some other interceptor etc. in the chain that will try to modify metadata, such that the singleton won't be appropriate?

is it possible to simply return nil?

File: pkg/config/tenancy/interceptor.go
Comment (yurishkuro) [inline pkg/config/tenancy/interceptor.go:1]: it's weird that we have interceptors and other functionality within package `config/*`. If config is only one of tenancy-related functions, then I would suggest:
```
pkg/
    tenancy/
        config.go (for TenancyConfig and validators)
        context.go (instead of being under storage/)
        grpc.go (for grpc-related pieces)
        http.go (for http-related pieces)
```

File: cmd/query/app/apiv3/grpc_gateway.go
Comment (yurishkuro) [inline cmd/query/app/apiv3/grpc_gateway.go:40]: why do we transform options to config here instead of doing once elsewhere and the reusing config across different handlers?

File: pkg/config/tenancy/context.go
Comment (esnible) [reply pkg/config/tenancy/context.go:75]: Can't pass `nil` because grpc-gateway's `runtime.WithMetadata()` expects a function that takes an `MD`, not a `*MD`.  With the current implementation of grpc-gateway returning a singleton isn't a problem, but there is no guarantee of that, and I will pass along a new MD.  This is only called if tenancy is enabled, and it will be a rare case that tenancy is enabled and clients aren't sending tenancy information.

File: storage/empty_test.go
Comment (yurishkuro) [inline storage/empty_test.go:1]: Did you mean to also delete `storage/tenant.go`?

File: cmd/query/app/server.go
Comment (yurishkuro) [inline cmd/query/app/server.go:164]: what about the API handler above, I assume it should also be wrapped in ExtractTenant

File: storage/empty_test.go
Comment (yurishkuro) [reply storage/empty_test.go:1]: +1

File: cmd/query/app/server.go
Comment (esnible) [reply cmd/query/app/server.go:172]: Thanks for your review.  *This comment was extremely helpful and identified several flaws.*

I have fixed the problems, but please don't merge yet! ;-)

The problems were not detected for two reasons.  I was testing with GetTrace, a streaming RPC, and ignoring tenancy for GetService, a unary RPC.  This PR needs unary RPC tests.  I have added one but need to increase coverage.
The other problem is that the tests set up the server using `newGRPCServer()` in cmd/query/app/grpc_handler_test.go, but the actual Jaeger uses `createGRPCServer()` which is nearly the same in implementation but was not tested for tenancy and didn't work.

I want to do at least one more commit of additional tests.

cc @pavolloffay

**Comment by esnible** [comment]
@pavolloffay I have added extensive tests.  Please re-review.

File: cmd/query/app/server.go
Comment (pavolloffay) [reply cmd/query/app/server.go:172]: I am not familiar with `newGRPCServer` nor `createGRPCServer` but if there is a way to unify these APIs we should consider doing it (e.g. in a follow up PR).

File: cmd/query/app/grpc_handler_test.go
Comment (pavolloffay) [inline cmd/query/app/grpc_handler_test.go:231]: Shouldn't be `tenancyConfig` passed instead of creating a new one?

**Review by pavolloffay — APPROVED** [review]

File: cmd/query/app/grpc_handler_test.go
Comment (esnible) [reply cmd/query/app/grpc_handler_test.go]: Now fixed, but of course that dismissed your review....

**Review by pavolloffay — APPROVED** [review]

**Review by pavolloffay — DISMISSED** [review]

File: pkg/tenancy/http.go
Comment (yurishkuro) [inline pkg/tenancy/http.go:28]: the logger is not used, let's remove it

File: cmd/query/app/http_handler.go
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:150]:
```suggestion
		handler = tenancy.ExtractTenantHTTPHandler(aH.tenancyConfig, handler, ah.logger)
```

File: cmd/query/app/handler_options.go
Comment (yurishkuro) [inline cmd/query/app/handler_options.go:66]: why pass options instead of tenancy.Config?

File: cmd/query/app/server.go
Comment (yurishkuro) [reply cmd/query/app/server.go:172]: > `tc := tenancy.NewTenancyConfig(&queryOpts.Tenancy)`

there are way too many places doing this conversion - we should only create tenancy config once (in the future it could be "live" object that self-updates).

File: pkg/tenancy/grpc.go
Comment (yurishkuro) [inline pkg/tenancy/grpc.go:89]: code below is identical to the previous function, please move to a single helper function like
```
tenant, err := GetAndValidateTenant(ctx)
if err != nil { return err }
if tenant == "" {
  return handler()
} else {
  return wrappedHandler()
}
```

File: cmd/query/app/server.go
Comment (esnible) [reply cmd/query/app/server.go:172]: @yurishkuro. I want to get this right.

I was using TenancyOptions for something human-readable, for Viper, and TenancyConfig for something machine readable, optimized for fast checking, following the pattern of `tlscfg.Options` and `tls.Config`.  Would it make sense to change tenency.TenancyConfig to `tenancy.TenancyManager`?

Currently the PR converts tenancy.Options to TenancyConfig in 4  non-test places.  Twice in `createGRPCServer()` -- I will fix that, once in `createHTTPServer()`, and once in `BuildHandlers()` (already merged; for trace ingestion).  Currently TenancyConfig is lightweight, and I didn't want to change the signatures to the create*Server() functions.  If we think of Tenancy is a "service" then I should create it once.  The next commit will address this.

Comment (yurishkuro) [reply cmd/query/app/server.go:172]: `TenancyConfig` is different from `tls.Config` in that the former is an object with behavior (an internal "service"), but the latter is just data passed elsewhere. I think from the naming it would be clearer if it was called `tenancy.TenancyManager`.

> Currently TenancyConfig is lightweight

It's only lightweight in the current implementation. Because it's a service, it can become arbitrarily complicated in the future, e.g. it could start refreshing the list of tenants from some storage, etc. By instantiating it in many places we introduce technical debt from the start that will need to be cleaned up later, so I'd rather fix it now and create this service once.

File: cmd/collector/app/span_handler_builder.go
Comment (yurishkuro) [inline cmd/collector/app/span_handler_builder.go:79]: we should treat TM similar to Logger - pass it to SpanHandlerBuilder at construction.

another peculiar thing: why is it `b.CollectorOpts.GRPC.Tenancy`? I.e. why tenancy options are within GRPC options?

File: cmd/query/app/http_handler.go
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:88]:
```suggestion
	tenancyMgr       *tenancy.TenancyManager
```

File: cmd/query/app/http_handler.go
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:117]: realistically, will we ever be in a situation where this is nil?  `cmd/query/app/server.go` does pass this option.

File: cmd/query/app/server.go
Comment (yurishkuro) [inline cmd/query/app/server.go:123]: you probably want `tm.Enabled` here

File: pkg/tenancy/config.go
Comment (yurishkuro) [inline pkg/tenancy/config.go:36]:
```suggestion
// NewTenancyManager creates TenancyManager from tenancy Options
```

File: pkg/tenancy/grpc_test.go
Comment (yurishkuro) [inline pkg/tenancy/grpc_test.go:28]: Is there a difference between these test cases and those in the next test?  Could we, perhaps, pull them out into a shared struct or function, and only differ in the business logic?

File: cmd/collector/app/handler/grpc_handler.go
Comment (esnible) [reply cmd/collector/app/handler/grpc_handler.go:65]: This line only became part of the PR because of the rename from `TenancyConfig` to `TenancyManager`.  It was already merged this way.  _oltp_receiver.go_:146 actually does pass in nil.  My thought at that time was to default to an empty handler.  If you would prefer it to be explicit I will also add an `error` return and fail if `nil` is passed in.  Thoughts?

File: cmd/collector/app/span_handler_builder.go
Comment (esnible) [reply cmd/collector/app/span_handler_builder.go:79]: This line only became part of the PR because of the rename from `TenancyConfig` to `TenancyManager`.  The collector doesn't use the interceptors, which were introduced for query.  (I tend to use interceptors as a last resort, as many devs struggle with them.)  Now that I introduced them for query it makes sense to use them for collection as well, and I will refactor this code to use them.

File: cmd/query/app/http_handler.go
Comment (esnible) [reply cmd/query/app/http_handler.go:117]: I am happy to remove the defaulting; I will add an error return at the same time.

File: cmd/collector/app/handler/grpc_handler.go
Comment (esnible) [reply cmd/collector/app/handler/grpc_handler.go:106]: This line only became part of the PR because of the rename from `TenancyConfig` to `TenancyManager`. The collector doesn't use the interceptors, which were introduced for query.  This PR is large; I propose backporting the new query interceptor to the collector should be in a follow-up.

Comment (yurishkuro) [reply cmd/collector/app/handler/grpc_handler.go:103]: +1

File: cmd/collector/app/collector.go
Comment (yurishkuro) [inline cmd/collector/app/collector.go:98]: when collector is instantiated from all-in-one, will this cause multiple tenancy managers to exist?

**Review by yurishkuro — COMMENTED** [review]
I think this PR is mostly ok to merge, but I would still prefer it is cleaned up to only instantiate tenancy manager in main's. I see at least two places where it's being created adhoc in the components that should instead be receiving it as a dependency.

**Comment by esnible** [comment]
@yurishkuro Thank you for your patience.  I was timid.  With the latest commit, only the _main.go_ and test cases create TenancyManager.  Thus there will now only be one of them even in the all-in-one case.

**Review by yurishkuro — APPROVED** [review]
🥳

**Pull request merged**
